### ISSUE TITLE
Meta-Learning is not out-of-the-box compatible with FLAX models which predict in batches.  Any ideas?

### ISSUE THREAD

**Issue by Defunctionalize** [google-deepmind/optax#449]
I'm sure this is something which is intuitively obvious to anyone who understands JAX's differentiation mechanism more deeply  😅 

I have re-created [the Meta Learning notebook](https://optax.readthedocs.io/en/latest/meta_learning.html) from Optax's docs, except instead of using a simple function to multiply by theta, I use an equivelant `flax.nn.Dense(features=1, use_bias=False)` model to generate the predictions.

the overall changes are very minor, [and viewable on this notebook](https://colab.research.google.com/drive/1mtLoFti89BJAmoO9xzynHgzlKX8-RWZI?usp=sharing).  The most relevant change is that instead of doing 

```
  state.hyperparams['learning_rate'] = jax.nn.sigmoid(eta)
  for x, y in samples[:-1]:
    theta, state = step(theta, state, x, y)
```
we simply do 

```
  state.hyperparams['learning_rate'] = jax.nn.sigmoid(eta)
  loss, theta, state= step(theta, state, x_batch, y_batch)
```
where loss is found with
```
  predictions = model.apply(theta, x_batch)
  return optax.l2_loss(predictions=predictions, targets=y_batch).mean()
```

As you can see in the notebook, this ends up not producing any gradient for the outer step to optimize.  Again, I'm sure this is something that would be immediately obvious to someone who understands how JAX creates gradients, but as this is much closer to the pattern used in a production system, and optimizing learning rate is such an important component of real world value, I thought it would be worth speaking out of ignorance to ask here --

does anyone have any ideas how one would could produce gradients while still operating on batches?

**Comment by jiriyu98** [comment]
Hello,

I'm learning JAX and am torn between using Flax or Haiku, happened to see this issue. I'm sure the author has figured it out, but it did confuse me at the begging, I would love to clarify it to benefit those who see it in the future. Also, correct me if I am wrong. ;-)

IIUC, the reproduction code didn't compute inner_loss, so instead of `return loss`. It should use `model_loss(theta, x_batch, y_batch)`. Or more intuitively in math, the `eta` is not related to any `loss` in this implementation, so that's why the gradient is always zero.

```
@jax.jit
def outer_loss(eta, theta, state, x_batch, y_batch):
  state.hyperparams['learning_rate'] = jax.nn.sigmoid(eta)
  loss, theta, state= step(theta, state, x_batch, y_batch)
  return loss, (theta, state)  # -> return model_loss(theta,  x_batch, y_batch), (theta, state) should work
```

BTW, I think JAX supports every pure function to get corresponding gradients and for `batch` case you mentioned, `jax.vmap` and `jax.pmap` might be a good choice.

Best,
